Fix logical error with what text is considered whitespace.

harrisonkaiser opened the pull request:
There appears to be a logical issue around what counts as leading white-space. There is code which does a subtraction assuming that no errors will be reported inside the leading whitespace. However we compute the length of that whitespace with std::char::is_whitespace and not rustc_lexer::is_whitespace. The former will include a no-break space while later will excluded it. We can only safely make the assumption that no errors will be reported  in whitespace if it is all "Rust Standard" whitespace. Indeed an error does occur in unicode whitespace if it contains a no-break space. In that case the subtraction will cause a ICE (for a compiler in debug mode) as described in https://github.com/rust-lang/rust/issues/132918.

uwaces committed:
Fix logical error with what text is considered whitespace.

There is a logical issue around what counts as leading white-space.
There is code which does a subtraction assuming that no errors will be reported
inside the leading whitespace. However we compute the length of
that whitespace with std::char::is_whitespace and not
rustc_lexer::is_whitespace. The former will include a no-break space while
later will excluded it. We can only safely make the assumption that no errors
will be reported  in whitespace if it is all "Rust Standard" whitespace.
Indeed an error does occur in unicode whitespace if it contains a no-break
space.

harrisonkaiser commented:
Addresses: https://github.com/rust-lang/rust/issues/132918#issuecomment-2544413808

jieyouxu commented on compiler/rustc_errors/src/emitter.rs:
```diff
@@ -1709,7 +1715,7 @@ impl HumanEmitter {
                                 }
                             })
                             .sum();
-                        if source_string.chars().any(|c| !c.is_whitespace()) {
+                        if source_string.chars().any(|c| rustc_lexer::is_whitespace(c)) {
```
Question: doesn't this flip the logic?

harrisonkaiser replied:
Yes. This is a typo.

jieyouxu reviewed (changes requested):
Thanks for the PR. We'll need to add a regression test for #132918 to prevent it from happening again in the future. Note that the test probably will need to be marked with

```
//@ needs-rustc-debug-assertions
```

to catch the overflow assertion that only happens if rustc was built with debug assertions.

Might also need to artificially constrain diagnostics width with something like

```
//@ compile-flags: --diagnostics-width=80
```

to prevent terminal size from interfering with test outcome. I'm not super familiar with what the logic here is intended to do, so I'm going to reroll a WG-diagnostics reviewer.

jieyouxu commented:
r? diagnostics

harrisonkaiser commented:
Thank you for the prompt response and guidance on adding a compilation tests.

I've fixed the typo and added a test case and confirmed that it doesn't pass on master and does pass on this branch.

harrisonkaiser commented:
> There are merge commits (commits with multiple parents) in your changes. We have a [no merge policy](https://rustc-dev-guide.rust-lang.org/git.html#no-merge-policy) so these commits will need to be removed for this pull request to be merged.
> 
> You can start a rebase with the following commands:
> 
> ```
> $ # rebase
> $ git pull --rebase https://github.com/rust-lang/rust.git master
> $ git push --force-with-lease
> ```
> 
> The following commits are merge commits:
> 
>     * [2d0fc53](https://github.com/rust-lang/rust/commit/2d0fc531141a6f1f4e395db464a81e5d06ec215f)

Whoops, fixed.

harrisonkaiser commented:
I noticed also that `is_whitespace` is also called here: (https://github.com/rust-lang/rust/blob/1e33dd17115ca948c6e5ebf319695ed64490b2bf/compiler/rustc_errors/src/emitter.rs#L789), and I'm don't know if that should also be `rustc_lexer::is_whitespace`. I figured I'd stick to the failure case I could re-produce.

harrisonkaiser commented:
@rustbot review

davidtwco commented:
> I noticed also that `is_whitespace` is also called here: (
> 
> https://github.com/rust-lang/rust/blob/1e33dd17115ca948c6e5ebf319695ed64490b2bf/compiler/rustc_errors/src/emitter.rs#L789
> 
> ), and I'm don't know if that should also be `rustc_lexer::is_whitespace`. I figured I'd stick to the failure case I could re-produce.

I'd be happy to see this changed to `rustc_lexer::is_whitespace`, it's probably better to be consistent. But that can be a follow-up so this can land quicker.

@bors r+ rollup

bors commented:
:pushpin: Commit 1e33dd17115ca948c6e5ebf319695ed64490b2bf has been approved by `davidtwco`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=davidtwco 1e33dd17115ca948c6e5ebf319695ed64490b2bf -->
<!-- homu: {"type":"Approved","sha":"1e33dd17115ca948c6e5ebf319695ed64490b2bf","approver":"davidtwco","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
